packages/design-system/src/components/TextField/TextField.jsx

```diff
@@ -22,110 +20,22 @@ export class TextField extends React.PureComponent {
-      this.focusRef && this.focusRef.focus();
-    }
-  }
-
-  /**
-   * @param {React.Component} field
-   * @returns {React.Component} The input field, optionally including mask
-   *  markup if a mask is present
-   */
-  renderFieldAndMask(field) {
-    return this.props.mask ? <Mask mask={this.props.mask}>{field}</Mask> : field;
-  }
-
   render() {
-    const {
-      ariaLabel,
-      className,
-      errorMessage,
-      fieldClassName,
-      focusTrigger,
-      hint,
-      id,
-      inversed,
-      inputRef,
-      label,
-      labelClassName,
-      labelId,
-      mask,
-      multiline,
-      numeric,
-      requirementLabel,
-      rows,
-      size,
-      type,
-      pattern,
-      ...fieldProps
-    } = this.props;
-    const FieldComponent = multiline ? 'textarea' : 'input';
-    const _rows = multiline && rows ? rows : undefined;
+    const containerProps = pick(this.props, containerPropList);
+    const inputProps = omit(this.props, containerPropList);
 
-    const classes = classNames(
+    // Add clearfix class
+    containerProps.className = classNames(
       'ds-u-clearfix', // fixes issue where the label's margin is collapsed
-      className
-    );
-
-    const fieldClasses = classNames(
-      'ds-c-field',
-      mask && `ds-c-field--${mask}`,
-      {
-        'ds-c-field--error': typeof errorMessage === 'string',
-        'ds-c-field--inverse': inversed,
-      },
-      fieldClassName,
-      size && `ds-c-field--${size}`
-    );
-
-    let inputType = type;
-    if (numeric) {
-      inputType = 'text';
-    } else if (multiline) {
-      inputType = undefined;
-    }
-
-    const field = (
-      <FieldComponent
-        aria-label={this.props.ariaLabel}
-        className={fieldClasses}
-        id={this.id}
-        /* eslint-disable no-return-assign */
-        ref={(ref) => {
-          if (focusTrigger) {
-            this.focusRef = ref;
-          } else {
-            if (inputRef) {
-              inputRef(ref);
-            }
-          }
-        }}
-        /* eslint-enable no-return-assign */
-        rows={_rows}
-        inputMode={numeric ? 'numeric' : undefined}
-        pattern={numeric && !pattern ? '[0-9]*' : pattern}
-        type={inputType}
-        {...fieldProps}
-      />
+      this.props.className
     );
 
     return (
-      <div className={classes}>
-        <FormLabel
-          className={labelClassName}
-          errorMessage={errorMessage}
-          fieldId={this.id}
-          hint={hint}
-          id={this.labelId}
-          requirementLabel={requirementLabel}
-          inversed={inversed}
-        >
-          {label}
-        </FormLabel>
-        {this.renderFieldAndMask(field, mask)}
-      </div>
+      <FieldContainer {...containerProps} component="div" labelComponent="label">
+        {({ className, errorMessage, id, inversed, name, setRef }) => (
+          <TextInput {...{ ...inputProps, className, errorMessage, id, inversed, name, setRef }} />
```

Comment:
@pwolfert Another alternative is for `FieldContainer` to only pass along transformed props like `id`, `labelId` and `setRef`. The other shared props still have to be handled in `TextField`, but this approach could be more explicit.

```
const containerProps = pick(this.props, containerPropList);
const inputProps = omit(this.props, omit(containerPropList, ['name', 'inversed', 'errorMessage']));

return (
    <FieldContainer {...containerProps}>
         {({  id, setRef }) => (
              <TextInput {...{ ...inputProps, id, setRef }} />
         )}
      </FieldContainer>
);
```

Reply:
I'm actually starting to like this approach a bit ^^
Especially if we remove the `fieldClassName` mapping logic from `FieldContainer`, this approach makes `FieldContainer` a bit simpler

Reply:
Is `containerPropList` equivalent to `Object.keys(FieldContainer.propTypes)`?

Reply:
Oh, one thing this pattern is not taking care of, though, is the `fieldClassNames` prop and any other ones you were transforming before

Reply:
I realize now that when I proposed
```js
<TextInput 
  {...pick(this.props, Object.keys(TextInput.propTypes))}
```
that doesn't account for the extra arbitrary props you can pass to the input like aria labels

Reply:
Even though the following with `errorMessage`, `inversed`, and `name` is more verbose than having it in an array, it's only a few special cases, and they get grouped visually with `className={this.props.fieldClassName}` which is also a special case
```js
const inputOnlyProps = omit(this.props, Object.keys(FieldContainer.propTypes));

return (
  <FieldContainer 
    {...this.props}
    className={containerClassName}
    component="div"
    labelComponent="label"
  >
    {({ id, setRef }) => (
      <TextInput 
        {...inputOnlyProps}
        {...{id, setRef}}
        className={this.props.fieldClassName}
        errorMessage={this.props.errorMessage}
        inversed={this.props.inversed}
        name={this.props.name}
      />
    )}
  </FieldContainer>
);
```
It may not be so bad to have special cases if it's really clear what they are, and I feel like it's pretty easy to read and see them when they're inlined in the JSX. You're welcome to disagree, of course. The difference I see here is that there is only one calculated `inputOnlyProps` at the top, and the rest is from explicit JSX prop assignments below.

Reply:
Yeah I agree, its only a few special cases. I also realized I can remove the `name` prop from `FieldContainer`, it was only used in generating ids anyway. 

I also realized, passing in  `this.props` into `FieldContainer` might not work with Typescript. I think I'll need to add a `pick(this.props, Object.keys(FieldContainer.propTypes));` to prevent undefined props errors. Still, since we can just use `FieldContainer.propTypes`, I feel this isn't so bad.

Reply:
Oh right, I was thinking arbitrary undefined props were fine since we did it on other components, but those "arbitrary" props are actually being inherited from `JSX.Element`, so they aren't actually arbitrary.

Reply:
Ah I forgot, this wont work either, because `fieldClassName` is still being passed to `TextInput` via `inputOnlyProps`.
```
    const containerProps = pick(this.props, Object.keys(FieldContainer.propTypes));
    const inputOnlyProps = omit(this.props, Object.keys(FieldContainer.propTypes));
    ...
   <TextInput
         {...inputOnlyProps}
         {...{ id, setRef }}
         className={this.props.fieldClassName}
```

I could do this:
```
const inputOnlyProps = omit(this.props, Object.keys(FieldContainer.propTypes).concat(['fieldClassName]);
```
But I'm thinking the extra complexity from prop renaming probably isnt worth it, especially since they are private components anyway. I'm going to just rename them to always be called `fieldClassName`

Reply:
I like the idea of just changing it to `fieldClassName` 👍